File: Leetcode Challenge/08_August_2020/Python/Week 4/3_Sum of Left Leaves.py
#Find the sum of all left leaves in a given binary tree.
#
#Example:
#
#    3
#   / \
#  9  20
#    /  \
#   15   7
#
#There are two left leaves in the binary tree, with values 9 and 15 respectively. Return 24.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumOfLeftLeaves(self, root: TreeNode) -> int:
        def aux(root, l):
            if not root:
                return 0
            else:
                if not root.left and not root.right:
                    return root.val * l
                else:
                    return aux(root.left, 1) + aux(root.right, 0)
        return aux(root, 0)

File: Leetcode Challenge/08_August_2020/C++/Week 4/3_Sum of Left Leaves.cpp
//Find the sum of all left leaves in a given binary tree.
//
//Example:
//
//    3
//   / \
//  9  20
//    /  \
//   15   7
//
//There are two left leaves in the binary tree, with values 9 and 15 respectively. Return 24.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumOfLeftLeaves(TreeNode* root) {
        return aux(root, 0);
    }
    
    int aux(TreeNode* root, int l) {
        if(root == nullptr) {
            return 0;
        } else {
            if(root->left == nullptr && root->right == nullptr) {
                return root->val * l;
            } else {
                return aux(root->left, 1) + aux(root->right, 0);
            }
        }
    }
};